helix-core/src/graphemes.rs

```diff
@@ -5,7 +5,79 @@ use ropey::{iter::Chunks, str_utils::byte_to_char_idx, RopeSlice};
 use unicode_segmentation::{GraphemeCursor, GraphemeIncomplete};
 use unicode_width::UnicodeWidthStr;
 
-use std::fmt;
+use std::borrow::Cow;
+use std::fmt::{self, Display};
+
+use crate::LineEnding;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// A preprossed Grapheme that is ready for rendering
+pub enum Grapheme<'a> {
+    Space,
+    Newline,
+    Nbsp,
+    Tab { width: u16 },
+    Other { raw: Cow<'a, str>, width: u16 },
+}
+
+impl<'a> Grapheme<'a> {
+    pub fn new(raw: Cow<'a, str>, visual_x: usize, tab_width: u16) -> Grapheme<'a> {
+        match &*raw {
+            "\t" => {
+                let width = tab_width - (visual_x % tab_width as usize) as u16;
+                Grapheme::Tab { width }
+            }
+            " " => Grapheme::Space,
+            "\u{00A0}" => Grapheme::Nbsp,
+            _ if LineEnding::from_str(&raw).is_some() => Grapheme::Newline,
+            _ => Grapheme::Other {
+                width: grapheme_width(&*raw) as u16,
+                raw,
+            },
+        }
+    }
+
+    pub fn change_position(&mut self, visual_x: usize, tab_width: u16) {
+        if let Grapheme::Tab { width } = self {
+            *width = tab_width - (visual_x % tab_width as usize) as u16
+        }
+    }
+
+    /// Returns the approximate visual width of this grapheme,
+    /// This serves as a lower bound for the width for use during soft wrapping.
+    /// The actual displayed witdth might be position dependent and larger (primarly tabs)
+    pub fn width(&self) -> u16 {
+        match *self {
+            Grapheme::Other { width, .. } | Grapheme::Tab { width } => width,
+            _ => 1,
+        }
+    }
+
+    pub fn is_whitespace(&self) -> bool {
+        !matches!(&self, Grapheme::Other { .. })
```

Comment:
This is only going to be true as long as the `Grapheme` enum variants other than `Other` only contains whitespace. Is this always going to be true? And further, is this much cheaper than e.g. just checking the Unicode properties?

Reply:
So there are a couple of reasons I used this 


* During rendering whitspace characters are configurable so it's useful to have them as separate variants
* Tabs are special in regards to their position calculation (width is position dependent) so I want them as an extra variant so I don't have to constantly recalculate the width
* newlines are special in that we want to treat them like whitespaces (even tough they could be a wide range of unicode codepoints) but when rendering I need to remember it was a whitespace so that we pick the correct whitespace char to display
* Detecting white-space is (or rather was) needed for soft-wrap

However I have changed to softwrap at word boundaries (see comment below) so you are right it doesn't make much sense anymore to keep special variants for spaces around.